### PR TITLE
gog-downloader: init at 1.2.3

### PR DESCRIPTION
**Pull request by RikudouSage**
## Description of changes

Adding a new package `gog-downloader` - a CLI tool to download game installers of games you own on GOG.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by keenanweaver** [comment]
Result of `nixpkgs-review pr 245971` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>gog-downloader</li>
  </ul>
</details>

File: maintainers/maintainer-list.nix
Comment (keenanweaver) [inline maintainers/maintainer-list.nix:1]: You may want to separate this into its own commit.

**Comment by keenanweaver** [comment]
I'm a longtime GOG fanboy (and former contractor for them), with a library of 3000 GOGs, so I'll be sure to try it out. Looks like nice software!

File: maintainers/maintainer-list.nix
Comment (RikudouSage) [reply maintainers/maintainer-list.nix:1]: In this one PR, or in separate PRs?

Comment (keenanweaver) [reply maintainers/maintainer-list.nix:1]: You can keep it in the same PR if you want to. Here's how I recently did it, if it helps: https://github.com/NixOS/nixpkgs/pull/239974/commits

Comment (RikudouSage) [reply maintainers/maintainer-list.nix:1]: Thanks, done!

**Review by keenanweaver — APPROVED** [review]
Looks good to me. However, I think it's preferred to have it build rather than retrieve from a compiled binary.

**Comment by RikudouSage** [comment]
Thanks! I will try, though it's gonna take me a while to figure it out, I'm really new to Nix and NixOS, this was kinda my first attempt at creating a package.

**Comment by keenanweaver** [comment]
I'm really new myself, so I don't really have the experience to provide any useful guidance. I see you have a CI workflow, basically just gotta Nix-ify that. I see there's a [PHP section in the nixpkgs manual](https://nixos.org/manual/nixpkgs/stable/#sec-php). Do you know of any other PHP apps that are in nixpkgs? I looked at the [php section](https://github.com/NixOS/nixpkgs/tree/master/pkgs/development/php-packages) and see a lot of them do the same you're already doing with fetchUrl, so maybe you're good to go already. Hoping someone with more experience chimes in.

File: pkgs/tools/games/gog-downloader/default.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+let
+  pname = "gog-downloader";
+  version = "1.2.3";
+  hash = "0kv09vmrz6dcjvaf2sw89cpb14wdr01r1kkzg84szfxwzr5bcc9p";
+in
+stdenv.mkDerivation {
+  inherit pname version hash;
+
+  src = fetchurl {
+    url = "https://github.com/RikudouSage/GogDownloader/releases/download/v${version}/gog-downloader";
```
Comment (drupol) [inline pkgs/tools/games/gog-downloader/default.nix:12]:
```suggestion
    url = "https://github.com/RikudouSage/GogDownloader/releases/download/v${finalAttrs.version}/gog-downloader";
```

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -41527,4 +41527,6 @@ with pkgs;
 
   wttrbar = callPackage ../applications/misc/wttrbar { };
 
+  gog-downloader = callPackage ../tools/games/gog-downloader { };
```
Comment (drupol) [inline pkgs/top-level/all-packages.nix:41530]: The letter `g` doesn't come after the letter `w`.

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution!

I've left some comments, left me know if this is clear enough.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (RikudouSage) [reply pkgs/top-level/all-packages.nix:41530]: Neither it does two lines before:

```
waylyrics = callPackage ../applications/audio/waylyrics { };

gitrs = callPackage ../tools/misc/gitrs { };
```

This doesn't seem to be sorted alphabetically at all, contrary to the maintainer list where I put it in alphabetical order.

File: pkgs/tools/games/gog-downloader/default.nix [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    sha256 = hash;
```
[the remark this answers is not in the document]
Comment (RikudouSage) [reply pkgs/tools/games/gog-downloader/default.nix:13]: Why? It works the same, but in my version all parameters that will change often are in one place.

File: pkgs/tools/games/gog-downloader/default.nix [hunk 4]
```diff
@@ -0,0 +1,40 @@
+{ stdenv, fetchurl, makeWrapper, lib, php }:
+
+let
+  pname = "gog-downloader";
+  version = "1.2.3";
+  hash = "0kv09vmrz6dcjvaf2sw89cpb14wdr01r1kkzg84szfxwzr5bcc9p";
+in
+stdenv.mkDerivation {
+  inherit pname version hash;
```
[the remark this answers is not in the document]
Comment (RikudouSage) [reply pkgs/tools/games/gog-downloader/default.nix:9]: Why? It was intentionally above to have all the often changeable configuration above everything else and visible at the first glance.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (drupol) [reply pkgs/top-level/all-packages.nix:41530]: Ok fair enough, let's leave it like that.

File: pkgs/tools/games/gog-downloader/default.nix [hunk 5]
```diff
@@ -0,0 +1,40 @@
+{ stdenv, fetchurl, makeWrapper, lib, php }:
```
[the remark this answers is not in the document]
Comment (RikudouSage) [reply pkgs/tools/games/gog-downloader/default.nix:1]: Is that a necessary change? If it's a common format in the repo, I'll be happy to change, even though I myself consider this syntax ugly.

I took my inspiration from [phpstan](https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/php-packages/phpstan/default.nix).

File: pkgs/tools/games/gog-downloader/default.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/tools/games/gog-downloader/default.nix:9]: I think we should use the `finalAttrs` pattern here, which is made to override everything properly.

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.

File: pkgs/tools/games/gog-downloader/default.nix (hunk 1 above)
Comment (RikudouSage) [reply pkgs/tools/games/gog-downloader/default.nix:12]: Do you have a link or anything (like a quick summary) where I can learn the difference? In the few packages I've seen I haven't noticed `finalAttrs` and I'd like to learn!

File: pkgs/tools/games/gog-downloader/default.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/tools/games/gog-downloader/default.nix:13]: I didn't say that "your version" was bad or not working. But I'm all in favor of consistency so, let's be consistent on how it's done on nixpkgs and do as I advised, for each comment in here.

File: pkgs/tools/games/gog-downloader/default.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/tools/games/gog-downloader/default.nix:1]: It's not necessary, but since we are reviewing, we know how things should be done in `nixpkgs`. Not everything is perfect, but we try our best to fix things piece by piece and avoid adding entropy in new PRs.

File: pkgs/tools/games/gog-downloader/default.nix (hunk 1 above)
Comment (drupol) [reply pkgs/tools/games/gog-downloader/default.nix:12]: See my comment in here: https://github.com/NixOS/nixpkgs/pull/245971#discussion_r1278619783

File: pkgs/tools/games/gog-downloader/default.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (RikudouSage) [reply pkgs/tools/games/gog-downloader/default.nix:13]: Yeah, consistency, sure. Side note, "do as I say because I said so" doesn't really help people understand your point.

[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/tools/games/gog-downloader/default.nix:13]: Feel free to let me know better wording here, I'm all ears.

[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/tools/games/gog-downloader/default.nix:13]: Prefer `hash = ...` to `sha256 = ...` Both currently work, but the former will make it easier if the algorithm ever needs to change.

Since there could be different hashing algorithms in use, this make sense.
